### ISSUE TITLE
How do make batchnorm to maintain multiple pairs of running_mean and running_var?

### ISSUE THREAD

**Issue by zhiqiangdon** [pytorch/pytorch#1738]
Hi,

The current batchnorm implementation only has one pair of running_mean and running_var. However, some modules in the neural networks may be used twice or more in one network forward. Based on my experimental experience and also stated in paper https://arxiv.org/pdf/1603.09025.pdf , different running_mean and running_var should be computed for different iterations(or time steps). I have found in my experiments, for the same testing data, net.eval() performs much worse than net.train() (backward is turned off here). The only difference in this situation is different running_mean and running_var are used. So how to make the batchnorm to keep multiple running_mean and running_var and I can choose to use the first pair in the first iteration, the second pair in the second iteration, and so on?

**Comment by soumith** [comment]
you have two separate BatchNorm layers, like b1 = nn.BatchNorm(...), b2 = nn.BatchNorm(...).

Also, questions are asked here: https://discuss.pytorch.org

**Comment by zhiqiangdon** [comment]
Thanks for the reply. The problem is that, as pointed in the paper mentioned above, the learnable parameters gamma and beta should be shared between these BN modules. Only the running_mean and running_var should be differentiated by the iterations. If do it in this way, how do we force them to share the learnable weights gamma and beta? @soumith
